### PR TITLE
feat: multi browser views support for BrowserWindow

### PR DESCRIPTION
**Pull request by summeroff**
#### Description of Change
new feature: multi browser views support for BrowserWindow 

Add functions addBrowserView, removeBroserView, getBrowserViews to BrowserWindow class. 

Existing API setBrowserView and getBrowserView changed to use new API inside.

This functionality was mentioned in the initial BrowserView feature PR https://github.com/electron/electron/pull/9166 . And asked few times since then. Recently in a discussion https://github.com/electron/electron/issues/10622 .

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes for `npm run test -match=BrowserView`, `all-test` run have few fails even on unchanged `2-0-x` branch.  
- [x] tests are [added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: multi browser views support for BrowserWindow

### CODE REVIEW

File: atom/browser/api/atom_api_window.cc [hunk 1]
```diff
@@ -914,6 +915,65 @@ void Window::ResetBrowserView() {
   browser_view_.Reset();
 }
 
+void Window::AddBrowserView(v8::Local<v8::Value> value) {
+  mate::Handle<BrowserView> browser_view;
+  if (value->IsNull() || value->IsUndefined()) {
```
Comment (zcbenz) [inline atom/browser/api/atom_api_window.cc]: There is no need to explicitly check `null`, type check can be down with:

```c++
if (mate::ConvertFromV8(isolate(), value, &browser_view)) {
} else {
  // Throw type error.
}
```

Comment (zcbenz) [inline atom/browser/api/atom_api_window.cc]: There is no need to check `null`, type check can be done with:

```c++
if (mate::ConvertFromV8(isolate(), value, &browser_view)) {
} else {
  // Throw type error.
}
```

`Window::RemoveBrowserView` can also be modified in the same way.

File: atom/browser/api/atom_api_window.cc [hunk 2]
```diff
@@ -914,6 +915,65 @@ void Window::ResetBrowserView() {
...
+
+      (*get_that_view).second.Reset(isolate(), value);
+      browser_views_.erase(get_that_view);
+    }
+    window_->UpdateDraggableRegionViews();
+  }
+}
+
+void Window::ResetBrowserViews() {
+  for (auto views_iter = browser_views_.begin();
```
Comment (zcbenz) [inline atom/browser/api/atom_api_window.cc]: Can use range-based for loop here:

```c++
for (auto item : browser_views_) {
  item.second.Reset()
}
```

Other for loops can also be modified as such.

File: atom/browser/native_window.h [hunk 3]
```diff
@@ -156,6 +156,8 @@ class NativeWindow : public base::SupportsUserData,
   virtual void SetMenu(AtomMenuModel* menu);
   virtual void SetParentWindow(NativeWindow* parent);
   virtual void SetBrowserView(NativeBrowserView* browser_view) = 0;
```
Comment (zcbenz) [inline atom/browser/native_window.h]: We can remove the `NativeWindow::SetBrowserView` implementation, and add a simple fallback for it by reusing the new API:

```js
BrowserWindow.prototype.setBrowserView = (view) => {
  this.resetBrowserViews()
  this.addBrowserView(view)
}
```

File: docs/api/browser-window.md [hunk 4]
```diff
@@ -1478,7 +1478,21 @@ removed in future Electron releases.
...
 #### `win.getBrowserView()` _Experimental_
 
-Returns `BrowserView | null` - an attached BrowserView. Returns `null` if none is attached.
+Returns `BrowserView | null` - an BrowserView attached by setBrowserView. Returns `null` if none is attached.
+
+#### `win.addBrowserView(browserView)` _Experimental_
+
+* `browserView` [BrowserView](browser-view.md)
+
+Replacement API for setBrowserView supporting work with multi browser views.
```
Comment (zcbenz) [inline docs/api/browser-window.md]: I think we should have a clear definition on the behavior when using `setBrowserView` and `addBrowserView` together, and have it documented. I suggest:

1. Make `setBrowserView` an alias for `removeAllBrowserViews() && addBrowserView(view)`.
2. Make `getBrowserView` throw if there are multiple browser views, and return `getBrowserViews()[0]` as fallback.

File: atom/browser/api/atom_api_window.cc (hunk 2 above)
Comment (summeroff) [reply atom/browser/api/atom_api_window.cc]: done.

File: atom/browser/native_window.h (hunk 3 above)
Comment (summeroff) [reply atom/browser/native_window.h]: I have tried to keep APIs separate to simplify usage in old code and new code written with new API. Even thinking about more explicit way what code can request new functionality. 
Your suggestion is good as it is let add new API without keeping legacy code and will not affect 
 existing user code. 
I have changed code in accordance with your suggestion.

File: atom/browser/api/atom_api_window.cc (hunk 1 above)
Comment (summeroff) [reply atom/browser/api/atom_api_window.cc]: Work of ConvertFromV8 is done by template implementation for corresponding type. For most types this indeed have check for null. But mate::Handle<> FromV8 converter does something opposite. 
```
 if (val->IsNull() || val->IsUndefined()) {
       *out = mate::Handle<T>();
       return true;
     }
```
And that lead to crash if null is passed. So I have to add check back. 
But change it to value->IsObject() 
Is it OK?

Do I have to add "Throw type error"? Prev API does not do it and as AddBrowserView works now as fall back for SetBrowserView. Then SetBrowserView start to throwing error too. But sending null to setBrowserView is required way to remove BrowserView what was set before.

File: docs/api/browser-window.md (hunk 4 above)
Comment (summeroff) [reply docs/api/browser-window.md]: documentation updated

**Comment by zcbenz** [comment]
Sorry I just noticed that this PR is targeted for 2.0.x. According to our versioning policy, we will not add new features to 2.0.x: https://www.electronjs.org/docs/tutorial/electron-versioning.

Do you mind rebasing your changes on `master` branch?

**Comment by summeroff** [comment]
Will it goes only to electron 4 from master or it can be backported to 3-0-x too?

**Comment by zcbenz** [comment]
New features will be included in 5.0, currently we don't have plans to publish minor updates for 3.x/4.x for new features.

**Pull request closed without merging**